glib-macros/src/enum_derive.rs

```diff
@@ -81,10 +100,228 @@ pub fn impl_enum(input: &syn::DeriveInput) -> TokenStream {
+                    match registration_status_ref_mut {
+                        // plugin has never been used (this is the first time), so postpones registration of the enum as a dynamic type on the first use.
+                        None => {
+                            *registration_status_ref_mut = Some((#crate_ident::clone::Downgrade::downgrade(type_plugin), #crate_ident::Type::INVALID));
+                            true
+                        },
+                        // plugin has been used at least one time and the enum has been registered as a dynamic type at least one time, so re-registers it.
+                        Some((_, type_)) if type_.is_valid() => {
+                            static mut VALUES: #enum_values_expr;
+                            *type_ = <#plugin_ty as glib::prelude::DynamicObjectRegisterExt>::register_dynamic_enum(type_plugin, #gtype_name, unsafe { &VALUES } );
+                            type_.is_valid()
+                        },
+                        // plugin has been used at least one time but the enum has not been registered yet as a dynamic type, so keeps postponed registration.
+                        Some(_) => {
+                            true
+                        }
+                    }
+                }
+
+                /// Depending on the plugin lifecycle state and on the registration status of the enum:
+                /// If plugin has been used (or reused) but the enum has not been registered yet as a dynamic type, cancels the postponed registration by deleting the weak reference on the plugin.
+                /// Else do nothing.
+                #[inline]
+                pub fn on_implementation_unload(type_plugin_: &#plugin_ty) -> bool {
+                    let registration_status_ref_mut = Self::get_registration_status_ref_mut();
+                    match registration_status_ref_mut {
+                        // plugin has never been used, so unload implementation is unexpected.
+                        None => false,
+                        // plugin has been used at least one time and the enum has been registered as a dynamic type at least one time.
+                        Some((_, type_)) if type_.is_valid() => true,
+                        // plugin has been used at least one time but the enum has not been registered yet as a dynamic type, so cancels the postponed registration.
+                        Some(_) => {
+                            *registration_status_ref_mut = None;
+                            true
+                        }
+                    }
+                }
+            }
+        }
+    } else {
+        // registers immediately the enum as a dynamic type.
+        quote! {
+            impl #name {
+                /// Returns a mutable reference to the glib type.
+                /// This is safe because the mutable reference guarantees that no other threads are concurrently accessing the atomic data.
```

Comment:
I don't think that's true. There's nothing preventing multiple threads from calling this function in parallel and each getting their own mutable reference.

This seems like it needs some changes

Reply:
Looks like a good approach, thanks :) Feel free to send the PR for fixing that in the other cases now already